packages/design-system/src/components/primitives/dnd/use-drop.ts

```diff
@@ -0,0 +1,293 @@
+        const toGlobalCoordinates = (placement: Placement | undefined) =>
+          placement && {
+            ...placement,
+            x: placement.x + parentRect.left,
+            y: placement.y + parentRect.top,
+          };
+
+        const neighbourRect = childrenRects[
+          indexAdjustment === 0 ? closestChildIndex - 1 : closestChildIndex + 1
+        ] as Rect | undefined;
+
+        const placement =
+          toGlobalCoordinates(
+            getPlacementBetween(closestChildRect, neighbourRect)
+          ) ||
+          toGlobalCoordinates(
+            getPlacementNextTo(
+              parentRect,
+              closestChildRect,
+              childrenOrientation,
+              indexAdjustment > 0 ? "forward" : "backward"
+            )
+          ) ||
+          getPlacementInside(parentRect, childrenOrientation);
+
+        const dropTarget: DropTarget<Data> = {
+          ...partialDropTarget,
+          rect: parentRect,
+          indexWithinChildren,
+          placement,
+        };
+
+        state.current.dropTarget = dropTarget;
+        latestProps.current.onDropTargetChange(dropTarget);
+      }
+    };
+
+    const detectTarget = () => {
+      const {
+        edgeDistanceThreshold = 3,
+        isDropTarget,
+        swapDropTarget,
+      } = latestProps.current;
+
+      const { pointerCoordinates } = state.current;
+      const root = rootRef.current;
+
+      // @todo: Cache this?
+      // Not expensive by itself, but it may call isDropTarget multiple times.
+      let candidate =
+        root &&
+        findClosestDropTarget({
+          root,
+          initialElement:
+            pointerCoordinates &&
+            toUseableElement(
+              document.elementFromPoint(
+                pointerCoordinates.x,
+                pointerCoordinates.y
+              )
+            ),
+          isDropTarget,
+        });
+
+      // To avoid calling swapDropTarget unnecessarily on every pointermove
+      const isNewCandidate =
+        candidate?.element !== state.current.lastCandidateElement;
+      state.current.lastCandidateElement = candidate?.element;
+      const candidateIsNearEdge =
+        candidate &&
+        pointerCoordinates &&
+        isNearEdge(
+          pointerCoordinates,
+          edgeDistanceThreshold,
+          candidate.element.getBoundingClientRect()
+        );
+      const isNewIsNearEdge =
+        candidateIsNearEdge !== state.current.lastCandidateIsNearEdge;
+      state.current.lastCandidateIsNearEdge = candidateIsNearEdge;
+      if (!isNewCandidate && !isNewIsNearEdge && state.current.dropTarget) {
+        // Still need to call setDropTarget to update rect and/or placement.
+        // Because indexWithinChildren might have changed,
+        // or parent coordinates might have moved in case of a scroll
+        setDropTarget(state.current.dropTarget);
+        return;
+      }
+
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
```

Comment:
this makes my heart stop every time I read it, I know in theory it should be fine, but maybe there is a better way to express it, e.g. with a condition of found candidate or something

Reply:
This wasn't the first choice for me either :)
But with `while(true)` the code is simplest compared to all other options

Reply:
I bet we can have an extra variable like foundSomething and while (foundSomething === undefined) or similar, would be one more line but 100% better